base/strings/annotated.jl

```diff
@@ -386,3 +386,51 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
 Get all annotations of `chr`.
 """
 annotations(c::AnnotatedChar) = c.annotations
+
+## AnnotatedIOBuffer
+
+struct AnnotatedIOBuffer <: IO
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+end
+
+AnnotatedIOBuffer(io::IOBuffer) = AnnotatedIOBuffer(io, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+AnnotatedIOBuffer() = AnnotatedIOBuffer(IOBuffer())
+
+function show(io::IO, annio::AnnotatedIOBuffer)
+    show(io, AnnotatedIOBuffer)
+    print(io, '(', annio.io.size, " bytes)")
+end
+
+position(io::AnnotatedIOBuffer) = position(io.io)
+lock(io::AnnotatedIOBuffer) = lock(io.io)
+unlock(io::AnnotatedIOBuffer) = unlock(io.io)
+
+function write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})
+    astr = AnnotatedString(astr)
+    offset = position(io.io)
+    for (region, annot) in astr.annotations
+        start, stop = first(region), last(region)
+        push!(io.annotations, (start+offset:stop+offset, annot))
+    end
+    write(io.io, astr)
+end
+write(io::AnnotatedIOBuffer, achr::AnnotatedChar) = write(io, AnnotatedString(achr))
+write(io::AnnotatedIOBuffer, x::AbstractString) = write(io.io, x)
+write(io::AnnotatedIOBuffer, s::Union{SubString{String}, String}) = write(io.io, x)
+write(io::AnnotatedIOBuffer, x::UInt8) = write(io.io, x)
+
+"""
+    read(io::AnnotatedIOBuffer, AnnotatedString)
+
+Read the entirety of `io`, as an `AnnotatedString`. This preserves the
+annotations of any `AnnotatedString`s written to `io` and otherwise acts like
+`read(io::IO, String)`.
+"""
+function read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{>:String}})
+    str = String(take!(io.io))
+    annots = copy(io.annotations)
+    empty!(io.annotations)
+    seekstart(io.io)
+    AnnotatedString(str, annotss)
+end
```

Comment:
I'm not super familiar with how to properly use the angry face emoji `>:`. Is the point here that this will be dispatched to on `read(::AnnotatedIOBuffer, AnnotatedString{String})` and `read(::AnnotatedIOBuffer, AnnotatedString{AbstractString})` and `read(::AnnotatedIOBuffer, AnnotatedString{Any})` but not for `read(::AnnotatedIOBuffer, AnnotatedString{MyString})` where `MyString <: AbstractString`?

Would this correctly support all parameterizations of AnnotateString:

```suggestion
function read(io::AnnotatedIOBuffer, ::Type{T}) where T <: AnnotatedString
    str = String(take!(io.io))
    annots = copy(io.annotations)
    empty!(io.annotations)
    seekstart(io.io)
    T(str, annotss)
end
```

Reply:
Well, we get an `AnnotatedString{String}` from this, so I think it would be reasonable to dispatch on all supertypes (of the `String` type parameter), which is really just:

```
julia> AnnotatedString{String} <: AnnotatedString{>:String}
true

julia> AnnotatedString{AbstractString} <: AnnotatedString{>:String}
true
```

However, this seems to prevent dispatch on the no-type-annotation form, so a `Union` might also be needed? I think for brevity we want `read(io, AnnotatedString)` to work.

Reply:
Oops, my initial suggestion was bad.
```suggestion
read(io::AnnotatedIOBuffer, ::Type{AnnotatedString}) = read(io, AnnotatedString{String})
function read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{T}}) where T
    str = read(io.io, T)
    annots = copy(io.annotations)
    empty!(io.annotations)
    seekstart(io.io)
    AnnotatedString(str, annotss)
end
```
IIUC, this would make an `AnnotatedString{T}` when passed a `AnnotatedString{T}` as input, provided you can read a `T` from the io object.

Reply:
I'm not sure how worthwhile this would be, given `T <: AbstractString` and that currently the only `T` for which this read method is implemented is `String`.

```

julia> methods(read, Tuple{IO, Type{<:AbstractString}})
# 4 methods for generic function "read" from Base:
 [1] read(stream, ::Type{Union{}}, slurp...; kwargs...)
     @ io.jl:222
 [2] read(io::IO, ::Type{T}) where T<:Enum
     @ Base.Enums Enums.jl:22
 [3] read(s::IO, ::Type{String})
     @ io.jl:1017
 [4] read(s::IO, T::Type)
     @ io.jl:1018
```

Reply:
The approach has been updated in subsequent conversation with Jamerson.